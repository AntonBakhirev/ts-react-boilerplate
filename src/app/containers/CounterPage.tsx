import * as React from "react";
import {connect} from "react-redux";
import {Dispatch} from "redux";
import {IStore} from "../redux/IStore";
import {
  decrement as decrementActionCreator,
  increment as incrementActionCreator
} from "../redux/modules/counterActionCreators";

interface IStateToProps {
  count: number;
}

interface IDispatchToProps {
  decrement: () => void;
  increment: () => void;
}

export interface IProps extends IStateToProps, IDispatchToProps {}

class CounterPage extends React.Component<IProps> {
  public render(): JSX.Element {
    const {count, decrement, increment} = this.props;
    return (
      <div>
        <h4>Counter</h4>
        <button name="incBtn" onClick={increment}>
          Increment
        </button>
        <button name="decBtn" onClick={decrement} disabled={count <= 0}>
          Decrement
        </button>
        <p>{count}</p>
      </div>
    );
  }
}

function mapStateToProps(state: Pick<IStore, "counter">): IStateToProps {
  return {
    count: state.counter.count
  };
}

function mapDispatchToProps(dispatch: Dispatch): IDispatchToProps {
  return {
    decrement: () => dispatch(decrementActionCreator()),
    increment: () => dispatch(incrementActionCreator())
  };
}

const connected = connect(mapStateToProps, mapDispatchToProps)(CounterPage);

export {connected as CounterPage, CounterPage as UnconnectedCounter, mapDispatchToProps, mapStateToProps};
