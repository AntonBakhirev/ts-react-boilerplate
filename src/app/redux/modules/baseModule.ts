import {Action} from "redux";

export interface IAction<P, T = string, M = string> extends Action<T> {
  payload?: P;
  message?: M;
}

export interface IBaseState {
  error: string;
  loaded: boolean;
  pending: boolean;
}

// trick compiler that they are literal types
export type INVOKED = "INVOKED";
export type PENDING = "PENDING";
export type FULFILLED = "FULFILLED";
export type REJECTED = "REJECTED";

export interface IPromiseActions {
  INVOKED: INVOKED;
  PENDING: PENDING;
  FULFILLED: FULFILLED;
  REJECTED: REJECTED;
}

export function getPromiseActionType(baseAction: string): IPromiseActions {
  return {
    FULFILLED: baseAction + "_FULFILLED" as FULFILLED,
    INVOKED: baseAction as INVOKED,
    PENDING: baseAction + "_PENDING" as PENDING,
    REJECTED: baseAction + "_REJECTED" as REJECTED
  };
}

export interface IActionCreators<I, P, F, R> {
  actionTypes: IPromiseActions;
  invoke: (payload: I) => IAction<I, INVOKED>;
  setPending: (payload: P) => IAction<P, PENDING>;
  setFulfilled: (payload: F) => IAction<F, FULFILLED>;
  setRejected: (payload: R) => IAction<null, REJECTED, R>;
}

export function getPromiseAction<I, P, F, R>(type: string): IActionCreators<I, P, F, R> {
  return {
    actionTypes: getPromiseActionType(type),
    invoke: (payload: I): IAction<I, INVOKED> => {
      return {type: type as INVOKED, payload};
    },
    setFulfilled: (payload: F): IAction<F, FULFILLED> => {
      return {type: getPromiseActionType(type).FULFILLED, payload};
    },
    setPending: (payload: P): IAction<P, PENDING> => {
      return {type: getPromiseActionType(type).PENDING, payload};
    },
    setRejected: (message: R): IAction<null, REJECTED, R> => {
      return {type: getPromiseActionType(type).REJECTED, message};
    }
  };
}
