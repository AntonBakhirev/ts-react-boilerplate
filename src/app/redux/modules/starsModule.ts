import {IStars, IStarsAction} from "models/starsModel";
import promiseReducer from "../../helpers/promiseReducer";

/** Action Types */
export const STARS_REQUEST: string = "stars/STARS_REQUEST";
export const STARS_SUCCESS: string = "stars/STARS_SUCCESS";
export const STARS_FAILURE: string = "stars/STARS_FAILURE";

/** Initial State */
const initialState: IStars = {
  isFetching: false
};

/** Reducer */
export function starsReducer(state: IStars = initialState, action: IStarsAction): IStars {
  return promiseReducer<IStars, IStarsAction>("stars/STARS", state, action);
}
