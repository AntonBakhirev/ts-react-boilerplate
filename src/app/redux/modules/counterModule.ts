import {IAction, IState} from "./baseModule";

export const INCREMENT: string = "counter/INCREMENT";
export const DECREMENT: string = "counter/DECREMENT";

export interface ICounter {
  count: number;
}

/** Counter: Initial State */
const initialState: IState<ICounter> = {
  isFetching: false,
  payload: {
    count: 0
  }
};

/** Reducer: CounterReducer */
export function counterReducer(state: IState<ICounter> = initialState, action?: IAction<ICounter>): IState<ICounter> {
  switch (action.type) {
    case INCREMENT:
      return {
        ...state,
        payload: {
          count: state.payload.count + 1
        }
      };

    case DECREMENT:
      return {
        ...state,
        payload: {
          count: ((state.payload.count - 1 > 0) ? state.payload.count - 1 : 0)
        }
      };

    default:
      return state;
  }
}
