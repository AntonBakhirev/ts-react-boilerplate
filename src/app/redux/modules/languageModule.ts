import promiseReducer, {IRequestType} from "../../helpers/promiseReducer";
import {IAction, IState} from "./baseModule";

export const SET_LANGUAGE = "language/SET_LANGUAGE";
export const SWITCH_LANGUAGE = "language/SWITCH_LANGUAGE";
export const requestType: IRequestType = {
  FAILURE: "language/FAILURE",
  PENDING: "language/PENDING",
  SUCCESS: SET_LANGUAGE
};

export interface ILanguage {
  languageData: any;
  locale: string;
}

const initialState: IState<ILanguage> = {
  isFetching: false,
  payload: {
    languageData: {},
    locale: "en-GB"
  }
};

export function languageReducer(state: IState<ILanguage> = initialState, action: IAction<ILanguage>): IState<ILanguage> {
  return promiseReducer<ILanguage>(requestType, state, action);
}
