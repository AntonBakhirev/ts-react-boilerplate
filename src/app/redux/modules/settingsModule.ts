import {FULFILLED, getPromiseAction, IAction, IBaseState, INVOKED, PENDING, REJECTED} from "./baseModule";

export const changeLanguage = getPromiseAction<string, null, any, string>("settings/CHANGE_LANGUAGE");

export interface ITranslations {
  [key: string]: string;
}

export interface ISettingsState extends IBaseState {
  language: string;
  translations: ITranslations;
}

const initialState: ISettingsState = {
  error: "",
  language: "en-US",
  pending: false,
  translations: {}
};

type CHANGE_LANGUAGE_PENDING = PENDING;
type CHANGE_LANGUAGE_FULFILLED = FULFILLED;
type CHANGE_LANGUAGE_REJECTED = REJECTED;

export function settingsReducer(
  state: ISettingsState = initialState,
  action: IAction<null, CHANGE_LANGUAGE_PENDING> |
    IAction<ITranslations, CHANGE_LANGUAGE_FULFILLED> |
    IAction<null, CHANGE_LANGUAGE_REJECTED, string>
): ISettingsState {
  switch (action.type) {
    case changeLanguage.actionTypes.PENDING:
      return {
        ...state,
        pending: true
      };
    case changeLanguage.actionTypes.FULFILLED:
      return {
        ...state,
        translations: action.payload
      };
    case changeLanguage.actionTypes.REJECTED:
      return {
        ...state,
        error: action.message,
        pending: false
      };
    default:
      return state;
  }
}

export function invokeChangeLanguage(language: string): IAction<string, INVOKED> {
  return {
    payload: language,
    type: changeLanguage.actionTypes.INVOKED
  };
}
