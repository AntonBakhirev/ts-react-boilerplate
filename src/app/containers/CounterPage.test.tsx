import {shallow} from "enzyme";
import * as React from "react";
import {IAction} from "../redux/modules/baseModule";
import {DECREMENT, ICounter, INCREMENT} from "../redux/modules/counterModule";
import {UnconnectedCounter} from "./CounterPage";

/* tslint:disable:no-empty jsx-no-lambda */
describe("<Counter />", () => {
  it("matches snapshot when count > 0", () => {
    const component = shallow(<UnconnectedCounter count={10} dispatch={jest.fn()} />);
    expect(component).toMatchSnapshot();
  });

  it("matches snapshot when count <= 0", () => {
    const component = shallow(<UnconnectedCounter count={0} dispatch={jest.fn()} />);
    expect(component).toMatchSnapshot();
  });

  it("calls handleIncrement() when increment button is clicked", () => {
    const spy = jest.spyOn(UnconnectedCounter.prototype, "handleIncrement");
    const shallowComponent = shallow(<UnconnectedCounter dispatch={() => {}} count={0}/>);

    expect(shallowComponent.find({name: "incBtn"})).toBeDefined();
    expect(spy).not.toHaveBeenCalled();
    shallowComponent.find({name: "incBtn"}).simulate("click");
    expect(spy).toHaveBeenCalled();
  });

  it("calls handleDecrement() when decrement button is clicked", () => {
    const spy = jest.spyOn(UnconnectedCounter.prototype, "handleDecrement");
    const shallowComponent = shallow(<UnconnectedCounter dispatch={() => {}} count={0}/>);

    expect(shallowComponent.find({name: "decBtn"})).toBeDefined();
    expect(spy).not.toHaveBeenCalled();
    shallowComponent.find({name: "decBtn"}).simulate("click");
    expect(spy).toHaveBeenCalled();
  });

  describe("handleIncrement()", () => {
    it("dispatches INCREMENT action", () => {
      const dispatch = jest.fn();
      const shallowComponent = shallow(<UnconnectedCounter dispatch={dispatch} count={0}/>);
      const expectedValue: IAction<ICounter> = {
        type: INCREMENT
      };

      expect(dispatch).not.toHaveBeenCalled();
      (shallowComponent as any).instance().handleIncrement();
      expect(dispatch).toHaveBeenCalledWith(expectedValue);
    });
  });

  describe("handleDecrement()", () => {
    it("dispatches DECREMENT action", () => {
      const dispatch = jest.fn();
      const shallowComponent = shallow(<UnconnectedCounter dispatch={dispatch} count={0}/>);
      const expectedValue: IAction<ICounter> = {
        type: DECREMENT
      };

      expect(dispatch).not.toHaveBeenCalled();
      (shallowComponent as any).instance().handleDecrement();
      expect(dispatch).toHaveBeenCalledWith(expectedValue);
    });
  });
});
