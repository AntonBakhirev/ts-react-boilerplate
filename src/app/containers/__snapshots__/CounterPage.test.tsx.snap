// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Counter /> matches snapshot when count <= 0 1`] = `
<div>
  <h4>
    <FormattedMessage
      defaultMessage="Counter"
      id="counter"
      values={Object {}}
    />
  </h4>
  <button
    name="incBtn"
    onClick={[Function]}
  >
    <FormattedMessage
      defaultMessage="Increment"
      id="counter.increment"
      values={Object {}}
    />
  </button>
  <button
    disabled={true}
    name="decBtn"
    onClick={[Function]}
  >
    <FormattedMessage
      defaultMessage="Decrement"
      id="counter.decrement"
      values={Object {}}
    />
  </button>
  <p>
    0
  </p>
</div>
`;

exports[`<Counter /> matches snapshot when count > 0 1`] = `
<div>
  <h4>
    <FormattedMessage
      defaultMessage="Counter"
      id="counter"
      values={Object {}}
    />
  </h4>
  <button
    name="incBtn"
    onClick={[Function]}
  >
    <FormattedMessage
      defaultMessage="Increment"
      id="counter.increment"
      values={Object {}}
    />
  </button>
  <button
    disabled={false}
    name="decBtn"
    onClick={[Function]}
  >
    <FormattedMessage
      defaultMessage="Decrement"
      id="counter.decrement"
      values={Object {}}
    />
  </button>
  <p>
    10
  </p>
</div>
`;
